File: Program_9/9(b).c
#include <stdio.h>
#include <limits.h>
struct Edge {
    int src, dest, weight;
};

int main() {

    int graph[6][6] = {
        {0, 4, 0, 0, 0, 2},
        {4, 0, 6, 0, 0, 3},
        {0, 6, 0, 3, 0, 0},
        {0, 0, 3, 0, 7, 0},
        {0, 0, 0, 7, 0, 1},
        {2, 3, 0, 0, 1, 0}
    };

    struct Edge edges[20];
    int E = 0;

    for (int i = 0; i < 6; i++) {
        for (int j = 0; j < 6; j++) {
            if (graph[i][j] != 0) {
                edges[E].src = i;
                edges[E].dest = j;
                edges[E].weight = graph[i][j];
                E++;
            }
        }
    }

    int dist[6];

    for (int i = 0; i < 6; i++)
        dist[i] = 99999;

    int src = 0;
    dist[src] = 0;

    for (int i = 1; i <= 6 - 1; i++) {
        for (int j = 0; j < E; j++) {
            int u = edges[j].src;
            int v = edges[j].dest;
            int w = edges[j].weight;

            if (dist[u] != 99999 && dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w;
            }
        }
    }

    for (int j = 0; j < E; j++) {
        int u = edges[j].src;
        int v = edges[j].dest;
        int w = edges[j].weight;

        if (dist[u] != 99999 && dist[u] + w < dist[v]) {
            printf("Graph contains a negative weight cycle!\n");
            return 0;
        }
    }
printf("Bellman-Ford Result (source = %d):\n", src);
    for (int i = 0; i < 6; i++)
        printf("Distance to %d = %d\n", i, dist[i]);

    return 0;
}

File: Program_9/9(b).py
class Edge():
    def __init__(self, src, dest, weight):
        self.src = src
        self.dest = dest
        self.weight = weight
    def printEdge(self):
        print(f"{self.src} -- {self.dest} == {self.weight}")

graph=[[0, 4, 0, 0, 0, 2],
       [4, 0, 6, 0, 0, 3],
       [0, 6, 0, 3, 0, 0],
       [0, 0, 3, 0, 7, 0],
       [0, 0, 0, 7, 0, 1],
       [2, 3, 0, 0, 1, 0]]

edges = []
E=0
for i in range(6):
    for j in range(6):
        if graph[i][j] != 0:
            edges.append(Edge(i, j, graph[i][j]))
            E+=1
dist=[9999]*6
src=0
dist[src]=0

for i in range(5):
    for j in range(E):
        u=edges[j].src
        v=edges[j].dest
        weight=edges[j].weight
        if dist[u] != 9999 and dist[u]+weight < dist[v]:
            dist[v]=dist[u]+weight

for i in range(E):
    u=edges[i].src
    v=edges[i].dest
    weight=edges[i].weight
    if dist[u] != 9999 and dist[u]+weight < dist[v]:
        print("Graph contains negative weight cycle")
        break

for i in range(6):
    print(f"Distance from source to vertex {i} is {dist[i]}")
    pass
for edge in edges:
    edge.printEdge()
print(dist)